### PR TITLE
Features/pharmacy api seutp refactor

### PR DESCRIPTION
**Pull request by Amuuo**
- added an `IPharmacyDbContext `
- refactored controller and service to handle more possible scenarios
- added `ServiceResult `class

### CODE REVIEW

File: PharmacyApi/Controllers/PharmacyController.cs [hunk 1]
```diff
@@ -26,27 +26,34 @@ public class PharmacyController : ControllerBase
 
     [HttpGet]
     [Route("all")]
-    public IActionResult GetAll() => Ok(_pharmacyService.GetAll());
+    public async Task<IActionResult> GetAll()
```
Comment (PixelPush3r) [inline PharmacyApi/Controllers/PharmacyController.cs]: Again - get what?

File: PharmacyApi/Services/PharmacyService.cs [hunk 2]
```diff
@@ -21,57 +22,115 @@ public class PharmacyService : IPharmacyService
...
-    public async Task<Pharmacy?> GetById(int id)
+    /// <summary>
+    /// Asynchronously retrieves a pharmacy record by its unique identifier.
+    /// </summary>
+    /// <param name="id">The unique identifier of the pharmacy record to retrieve.</param>
+    /// <returns>
+    /// A ServiceResult containing the Pharmacy object if found, 
+    /// or an error message if no pharmacy record with the specified id exists.
+    /// </returns>
+    public async Task<ServiceResult<Pharmacy>> GetByIdAsync(int id)
```
Comment (PixelPush3r) [inline PharmacyApi/Services/PharmacyService.cs]: Get what?
Also is by id really normal? most folks know

File: PharmacyApi/Controllers/PharmacyController.cs [hunk 3]
```diff
@@ -24,29 +24,34 @@ public class PharmacyController : ControllerBase
...
+            var pharmacyListResult = await _pharmacyService.GetPharmacyListAsync();
+            return pharmacyListResult.Success 
+                ? Ok(pharmacyListResult.Result) 
+                : NotFound(pharmacyListResult.ErrorMessage);
+        }
+
+        var pharmacyResult = await _pharmacyService.GetByIdAsync(id.Value);
+        return pharmacyResult.Success 
+            ? Ok(pharmacyResult.Result) 
+            : NotFound(pharmacyResult.ErrorMessage);
```
Comment (PixelPush3r) [inline PharmacyApi/Controllers/PharmacyController.cs]: Why stop combining the calls at this level?
I see duplicate code... why not combine the operation on the Service layer?

File: PharmacyApi/Controllers/PharmacyController.cs [hunk 4]
```diff
@@ -24,29 +24,34 @@ public class PharmacyController : ControllerBase
...
+    
 
     [HttpPost]
-    [Route("update/{id}")]
-    public async Task<IActionResult> UpdateById(int id, Pharmacy pharmacy)
+    [Route("update")]
+    public async Task<IActionResult> UpdateById(Pharmacy pharmacy)
     {
-        var updatedPharmacy = await _pharmacyService.UpdateById(id, pharmacy);
+        var updatedPharmacyResult = await _pharmacyService.UpdateByIdAsync(pharmacy);
```
Comment (PixelPush3r) [inline PharmacyApi/Controllers/PharmacyController.cs]: Good!

File: PharmacyApi/Controllers/PharmacyController.cs [hunk 5]
```diff
@@ -24,29 +24,34 @@ public class PharmacyController : ControllerBase
...
 
-        return updatedPharmacy is not null
-            ? Ok(updatedPharmacy)
-            : StatusCode(404, $"No pharmacy found with id {id}");
+        return updatedPharmacyResult.Success
+            ? Ok(updatedPharmacyResult.Result)
+            : NotFound(updatedPharmacyResult.ErrorMessage);
```
Comment (PixelPush3r) [inline PharmacyApi/Controllers/PharmacyController.cs]: Better, but NotFound means "404 - Page not found" not "Pharmacy Not Found..."
Another reason why i dislike RestFulness... twists the meaning up...
When you cant find an entity - that should be a different error code...

File: PharmacyApi/Data/PharmacyDbContext.cs [hunk 6]
```diff
@@ -1,16 +1,21 @@
...
+    private readonly ILogger _logger;
 
+    public PharmacyDbContext(DbContextOptions options, 
+                             ILogger<PharmacyDbContext> logger) : base(options)
+    {
+        _logger = logger;
     }
 
-    public DbSet<Pharmacy> Pharmacies { get; set; }
+    public DbSet<Pharmacy> PharmacyList { get; set; }
```
Comment (PixelPush3r) [inline PharmacyApi/Data/PharmacyDbContext.cs:14]: Good!

File: PharmacyApi/Program.cs [hunk 7]
```diff
@@ -32,12 +28,11 @@
     options.UseSqlServer(connectionString);
 });
 
-builder.Services.AddTransient<IPharmacyService, PharmacyService>();
+builder.Services.AddScoped<IPharmacyService, PharmacyService>();
+builder.Services.AddScoped<IPharmacyDbContext, PharmacyDbContext>();
```
Comment (PixelPush3r) [inline PharmacyApi/Program.cs]: Scoped - you will have problems when more than 4 people hit your api at the same time...
It takes a while to new-up a DbContext... then again is your context structure going to change at runtime?... nope...

File: PharmacyApi/Services/IPharmacyService.cs [hunk 8]
```diff
@@ -1,11 +1,12 @@
...
+using PharmacyApi.Utilities;
 
 namespace PharmacyApi.Services;
 
 public interface IPharmacyService
 {
-    IEnumerable<Pharmacy> GetAll();
-    Task<Pharmacy?> GetById(int id);
-    Task<Pharmacy?> UpdateById(int id, Pharmacy updatedPharmacy);
+    Task<ServiceResult<IAsyncEnumerable<Pharmacy>>> GetPharmacyListAsync();
```
Comment (PixelPush3r) [inline PharmacyApi/Services/IPharmacyService.cs:8]: "IAsyncEnumerable" - thats an interesting interface...

File: PharmacyApi/Services/IPharmacyService.cs [hunk 9]
```diff
@@ -1,11 +1,12 @@
...
+    Task<ServiceResult<Pharmacy>> GetByIdAsync(int id);
+    Task<ServiceResult<Pharmacy>> UpdateByIdAsync(Pharmacy updatedPharmacy);
```
Comment (PixelPush3r) [inline PharmacyApi/Services/IPharmacyService.cs]: Dont name methods with generic names - unless you are supplying a truly generic function - which this is definitely not...
Use the Pharmacy in the name all the way up and down.

File: PharmacyApi/Utilities/ServiceResult.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿namespace PharmacyApi.Utilities
+{
+    public class ServiceResult<T>
+    {
+        public T? Result { get; set; }
+        public bool Success { get; set; }
+        public string? ErrorMessage { get; set; }
+    }
```
Comment (PixelPush3r) [inline PharmacyApi/Utilities/ServiceResult.cs]: Interesting - this is a good approach. Can it be enriched w error / success in addition to the overall success flag?

File: PharmacyApi/Utilities/ServiceResult.cs [hunk 11]
```diff
@@ -0,0 +1,9 @@
+﻿namespace PharmacyApi.Utilities
+{
+    public class ServiceResult<T>
+    {
+        public T? Result { get; set; }
+        public bool Success { get; set; }
```
Comment (PixelPush3r) [inline PharmacyApi/Utilities/ServiceResult.cs]: Boolean - Specifying Boolean field names as questions is best practice: "IsSuccess"
Some old timers like to call it "SuccessFlag" which is fine, if you still use a washboard and clothesline in your backyard...

**Review by PixelPush3r — CHANGES_REQUESTED** [review]

File: PharmacyApi/Utilities/ServiceResult.cs (hunk 10 above)
Comment (Amuuo) [reply PharmacyApi/Utilities/ServiceResult.cs]: How do you mean?

Comment (PixelPush3r) [reply PharmacyApi/Utilities/ServiceResult.cs]: Nice to know if its a 500 system error vs a handled Unsuccessful / BadRequest error...

File: PharmacyApi/Controllers/PharmacyController.cs (hunk 5 above)
Comment (Amuuo) [reply PharmacyApi/Controllers/PharmacyController.cs]: How could I make this more elegant?

**Pull request merged**